wings/market/coinbase_pro_market.pyx

```diff
@@ -0,0 +1,816 @@
+            'CB-ACCESS-TIMESTAMP': timestamp,
+            'CB-ACCESS-KEY': self.api_key,
+            'CB-ACCESS-PASSPHRASE': self.passphrase,
+            'Content-Type': 'application/json',
+        }
+
+
+cdef class InFlightOrder:
+    cdef:
+        public str client_order_id
+        public str exchange_order_id
+        public str symbol
+        public bint is_buy
+        public object order_type
+        public object amount
+        public object price
+        public object executed_amount
+        public object quote_asset_amount
+        public str fee_asset
+        public object fee_paid
+        public str last_state
+        public object exchange_order_id_update_event
+
+    def __init__(self,
+                 client_order_id: str,
+                 exchange_order_id: str,
+                 symbol: str,
+                 is_buy: bool,
+                 order_type: OrderType,
+                 amount: Decimal,
+                 price: Decimal):
+        global s_decimal_0
+
+        self.client_order_id = client_order_id
+        self.exchange_order_id = exchange_order_id
+        self.symbol = symbol
+        self.is_buy = is_buy
+        self.order_type = order_type
+        self.amount = amount
+        self.price = price
+        self.executed_amount = s_decimal_0
+        self.quote_asset_amount = s_decimal_0
+        self.fee_asset = None
+        self.fee_paid = s_decimal_0
+        self.last_state = 'open'
+        self.exchange_order_id_update_event = asyncio.Event()
+
+    def __repr__(self) -> str:
+        return f"InFlightOrder(client_order_id='{self.client_order_id}', exchange_order_id={self.exchange_order_id}, " \
+               f"symbol='{self.symbol}', is_buy={self.is_buy}, order_type={self.order_type_description}, " \
+               f"amount={self.amount}, price={self.price}, executed_amount={self.executed_amount}, "\
+               f"quote_asset_amount={self.quote_asset_amount}, fee_asset='{self.fee_asset}', "\
+               f"fee_paid={self.fee_paid}, last_state='{self.last_state}')"
+
+    @property
+    def is_done(self) -> bool:
+        return self.last_state == "done"
+
+    @property
+    def is_failure(self) -> bool:
+        return self.last_state not in {"open", "active", "pending", "done"}
+
+    @property
+    def base_asset(self) -> str:
+        return self.symbol.split("-")[0]
+
+    @property
+    def quote_asset(self) -> str:
+        return self.symbol.split("-")[1]
+
+    @property
+    def order_type_description(self) -> str:
+        order_type = "market" if self.order_type is OrderType.MARKET else "limit"
+        side = "buy" if self.is_buy else "sell"
+        return f"{order_type}{side}"
+
+    def update_exchange_order_id(self, exchange_id: str):
+        self.exchange_order_id = exchange_id
+        self.exchange_order_id_update_event.set()
+
+    async def get_exchange_order_id(self):
+        if len(self.exchange_order_id) == 0 or self.exchange_order_id is None:
```

Comment:
this check seems strange, can the order id be empty string?

Reply:
nm i see that tracked orders are initialized with "". maybe just check if its == "" directly

Reply:
changing to `if self.exchange_order_id == "":`